content/blog/next-level-iac-breakpoint-debugging/index.md

```diff
@@ -0,0 +1,342 @@
+---
+title: "Next-level IaC: Breakpoint Debugging for Pulumi Programs"
+date: 2024-05-31
+draft: false
+meta_desc: "Next-level IaC: Breakpoint Debugging for Pulumi Programs"
+meta_image: meta.png
+authors:
+    - troy-howard
+tags:
+    - next-level-iac
+    - debugging
+    - vscode
+    - typescript
+    - nodejs
+    - ide
+---
+
+For many of us, life is suspended between presses of the [`F5`][f5-key] key. Our IDE is the only place where everything is organized and makes sense. And while we know that the likelihood of a code hole-in-one is very rare (that magical moment where you run your build and everything *just works* the first time), we always hope that *this* time, this will be the run where you see the output that lets you know you've gotten it right. And in between? Debugging.
+
+<!--more-->
+
+Debugging is a [complicated topic][debugging]. But what is not complicated is understanding [the cost of debugging][cost-of-debugging]. Debugging is the biggest unknown-unknown in software. It makes you miss your sprint goals, it holds back scaling and company growth, it slows your time to market. Debugging is what makes estimating so difficult, because until you know what the problem is, you don't know how long it will take to fix, or indeed if it is even possible to fix at all. Which is why savvy developers invest a lot of time into refining their debugging process and seek out the best-in-class tools to streamline this process.
+
+In the infrastructure world, debugging is a pretty big problem. Some of the most popular tools have nothing more than simple unstructured text output logs. While tools like Terraform call themselves "Infrastructure as Code" the "code" is a proprietary DSL, not a general purpose programming language, and so there is little-to-no IDE support. There is no `F5`. Even worse, there is no `F9` for setting a breakpoint. There's no way to inspect state mid-run, or more importantly to explore in an unstructured way, because this is where the "aha!" moments usually happen.
+
+Pulumi gives you a rich IaC debugging experience by enabling breakpoint debugging in your favorite IDE. Let's walk through an example of how to enable a full-featured debugging workflow in [VS Code][vs-code] for TypeScript.
+
+## How Breakpoint Debugging Works
```

Comment:
should this section be put in one of these

`{{% notes type="info" %}}`

And then hidden behind a expander? @cnunciato what do you think?